# 06 - Classificação manual

Qual é o resultado para a previsão baseada na famosa frase: "mulheres e crianças primeiro"?

## Preparando o ambiente

In [5]:
import pandas as pd
import numpy as np
from sklearn.metrics import accuracy_score

## Carregando os dados

In [2]:
titanic = pd.read_csv('https://raw.githubusercontent.com/SalatielBairros/kaggle-titanic/main/data/processed/train_processed_byage_ag.csv')
del titanic['Unnamed: 0']
titanic.head(5)

Survived  Pclass   Sex  Age  SibSp  Parch Floor  Relateds  faixa_etaria  \
0         0       3  male   22      1      0    SC         1  jovem_adulto   
1         0       3  male   22      0      0    SC         0  jovem_adulto   
2         0       3  male   22      0      0    SC         0  jovem_adulto   
3         0       3  male   22      0      0    SC         0  jovem_adulto   
4         0       3  male   22      0      0    SC         0  jovem_adulto   

   acompanhado  possui_cabine  
0         True          False  
1        False          False  
2        False          False  
3        False          False  
4        False          False

In [3]:
teste = pd.read_csv('https://raw.githubusercontent.com/SalatielBairros/kaggle-titanic/main/data/processed/test_processed.csv')
del teste['Unnamed: 0']
Xvalidacao = teste.drop(columns=['PassengerId'])
teste.head(5)

PassengerId  Pclass     Sex  Age  SibSp  Parch Floor  Relateds  \
0          892       3    male   34      0      0    SC         0   
1          893       3  female   47      1      0    SC         1   
2          894       2    male   62      0      0    SC         0   
3          895       3    male   27      0      0    SC         0   
4          896       3  female   22      1      1    SC         2   

   possui_cabine  acompanhado  faixa_etaria  
0          False        False  jovem_adulto  
1          False         True  adulto_idoso  
2          False        False  adulto_idoso  
3          False        False  jovem_adulto  
4          False         True  jovem_adulto

In [7]:
X = titanic.drop(columns=['Survived'])
y = titanic['Survived']

## Criando o algoritmo de classificação manual

In [20]:
def predict(df):
    predicoes = []
    for index, row in df.iterrows():
        if(row['Sex'] == 'female' or row['faixa_etaria'] == 'crianca_adolescente'):
            predicoes.append(1)
        else:
            predicoes.append(0)
    return predicoes

In [22]:
accuracy_score(y, predict(X))

0.7890011223344556

A acurácia baseada nessa regra foi de aproximadamente **78.9%** para os dados de treino. E para os dados de teste?

In [23]:
predicoes = predict(Xvalidacao)
resultado = pd.DataFrame()
resultado['PassengerId'] = teste['PassengerId']
resultado['Survived'] = predicoes
resultado.sample(5)

PassengerId  Survived
163         1055         0
385         1277         1
49           941         1
205         1097         0
352         1244         0

In [24]:
resultado.to_csv('../../data/submissions/manual.csv', index=False)

## Resultado

O resultado obtido no Kaggle foi de **75.3%** de acurácia. Bem semelhante ao obtido com os dados de treinamento. Dessa forma é possível estipular que este seja o valor mínimo, visto que é maior que o do `DummyClassifier`.